File: ml-service/pipelines/interviewer_briefing.py
"""Interviewer briefing — great-harness-agent offers.py Sub-Agent 2 style."""

from pipelines.groq_service import GroqApiError, groq_strong, require_groq


def _answers_from_transcript(transcript: list) -> str:
    answers = [
        m.get("text", "")
        for m in (transcript or [])
        if m.get("speaker") in ("You", "Candidate")
    ]
    return "\n".join(f"- {a}" for a in answers[:8] if str(a).strip())


def generate_interviewer_briefing(payload: dict) -> dict:
    """Generate detailed interviewer briefing with screening + AI interview insights."""
    candidate_name = payload.get("candidate_name") or "Candidate"
    job_title = payload.get("job_title") or "Role"
    interviewer_name = payload.get("interviewer_name") or "Interviewer"
    interviewer_role = payload.get("interviewer_role") or "Panel Member"

    screening = payload.get("screening") or {}
    interview = payload.get("interview") or {}
    app = payload.get("application") or {}

    interview_score = interview.get("interview_score") or interview.get("interviewScore") or "N/A"
    verdict = interview.get("verdict") or "N/A"
    composite = interview.get("composite_score") or interview.get("compositeScore")
    screening_score = (
        screening.get("total_score")
        or screening.get("ai_score")
        or app.get("jd_score")
        or app.get("jdScore")
        or "N/A"
    )

    strengths = (
        screening.get("top_strengths")
        or app.get("matched_skills")
        or app.get("matchedSkills")
        or interview.get("top_strengths")
        or interview.get("topStrengths")
        or []
    )
    concerns = (
        interview.get("concerns")
        or screening.get("key_gaps")
        or app.get("missing_skills")
        or app.get("missingSkills")
        or []
    )
    recommendation = interview.get("recommendation") or screening.get("verdict") or ""
    jd_summary = app.get("jd_fit_summary") or app.get("jdFitSummary") or screening.get("decision_note") or ""

    transcript = interview.get("harness_transcript") or interview.get("harnessTranscript") or []
    answers_summary = _answers_from_transcript(transcript)
    if not answers_summary:
        answers_summary = "\n".join(
            f"- {str(a.get('answer') or a.get('transcript') or '')[:200]}"
            for a in (interview.get("answers") or interview.get("qa_log") or [])[:8]
            if str(a.get("answer") or a.get("transcript") or "").strip()
        )

    dim_lines = []
    dim_map = [
        ("Technical", ("technical_score", "technicalScore")),
        ("Communication", ("communication_score", "communicationScore")),
        ("Problem solving", ("problem_solving_score", "problemSolvingScore")),
        ("Culture fit", ("culture_fit_score", "cultureFitScore")),
        ("Experience depth", ("experience_depth_score", "experienceDepthScore")),
        ("JD alignment", ("jd_alignment_score", "jdAlignmentScore")),
    ]
    for label, keys in dim_map:
        val = next((interview.get(k) for k in keys if interview.get(k) is not None), None)
        if val is not None:
            dim_lines.append(f"- {label}: {round(float(val))}/100")

    prompt = f"""You are preparing a briefing for {interviewer_name} ({interviewer_role}) who will conduct a final technical interview with {candidate_name} for a {job_title} role.

=== RESUME SCREENING RESULTS ===
Score: {screening_score}/100
Verdict: {screening.get('verdict') or app.get('recommendation') or 'Reviewed'}
Summary: {jd_summary}
Matched skills: {', '.join(strengths[:12]) if strengths else 'N/A'}
Gaps / missing: {', '.join(concerns[:12]) if concerns else 'N/A'}
Red flags: {', '.join(screening.get('red_flags') or []) or 'None noted'}

=== AI VOICE INTERVIEW RESULTS ===
Interview score: {interview_score}/100
Verdict: {verdict}
Composite (80% screening + 20% interview): {composite if composite is not None else 'N/A'}/100
Recommendation: {recommendation}
AI feedback: {interview.get('ai_feedback') or interview.get('aiFeedback') or 'N/A'}
Dimension scores:
{chr(10).join(dim_lines) if dim_lines else '- Not available'}

=== CANDIDATE'S KEY ANSWERS FROM AI INTERVIEW ===
{answers_summary or 'No transcript captured.'}

=== GENERATE BRIEFING ===

Write an HTML-friendly briefing (use <p>, <ul>, <ol>, <strong> — no outer <html> wrapper) with these sections:

1. CANDIDATE SNAPSHOT (3-4 sentences)
   Who they are, experience level, tech stack, standout signals.

2. RESUME SCREENING HIGHLIGHTS
   - Strong areas (3 bullets)
   - Areas needing verification (3 bullets)
   - Red flags if any (1-2 bullets)

3. AI INTERVIEW — HOW THEY PASSED
   Scores, verdict, strengths, concerns, manager summary.

4. SUGGESTED INTERVIEW QUESTIONS (8-10 questions)
   Deep-dive on projects, probe screening gaps, live coding, system design.
   Format: numbered list with brief "what to look for" under each.

5. EVALUATION CRITERIA
   Rate after session: Technical Depth, Problem Solving, Communication, Culture Fit (each 1-5).

Keep it concise, actionable, and specific to this candidate."""

    require_groq()
    text = groq_strong(
        "You are an expert technical hiring manager preparing interview briefings. "
        "Use ONLY the candidate data provided — no generic filler.",
        prompt,
    )
    if not text or not str(text).strip():
        raise GroqApiError("Groq did not return interviewer briefing content.")
    return {"briefing_html": str(text).strip(), "generated_by": "groq"}

File: backend-express/src/lib/interviewerBriefing.js
/** Interviewer briefing — great-harness-agent offers.py Sub-Agent 2 style. */

const ml = require('../services/mlClient');

function esc(s) {
  return String(s || '').replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;');
}

function listItems(items) {
  if (!items?.length) return '<p><em>None noted</em></p>';
  return `<ul>${items.map((i) => `<li>${esc(i)}</li>`).join('')}</ul>`;
}

function extractCandidateAnswers(interview) {
  const fromTranscript = (interview?.harnessTranscript || [])
    .filter((m) => m.speaker === 'You' || m.speaker === 'Candidate')
    .map((m) => m.text)
    .filter(Boolean);
  if (fromTranscript.length) return fromTranscript.slice(0, 8);

  return (interview?.qaLog || interview?.answers || [])
    .map((a) => a.answer || a.transcript || '')
    .filter((t) => String(t).trim().length > 10)
    .slice(0, 8);
}

function buildSuggestedQuestions(app, interview) {
  const fromInterview = (interview?.questions || [])
    .slice(0, 4)
    .map((q) => (typeof q === 'string' ? q : q.question))
    .filter(Boolean);
  const gaps = app.screening?.key_gaps || app.missingSkills || [];
  const gapQs = gaps.slice(0, 3).map((g) => `Probe depth on: ${g}`);
  const projectQ = app.parsedData?.experience?.[0]
    ? `Deep-dive: ${app.parsedData.experience[0].title || 'recent role'} at ${app.parsedData.experience[0].company || 'previous company'}`
    : 'Walk through your most complex recent project end-to-end';
  return [projectQ, ...fromInterview, ...gapQs].slice(0, 10);
}

function dimensionRows(interview) {
  const rows = [
    ['Technical depth', interview?.technicalScore],
    ['Communication', interview?.communicationScore],
    ['Problem solving', interview?.problemSolvingScore],
    ['Culture fit', interview?.cultureFitScore],
    ['Experience depth', interview?.experienceDepthScore],
    ['JD alignment', interview?.jdAlignmentScore],
  ];
  return rows.filter(([, v]) => v != null && v > 0);
}

function buildStaticInterviewerBriefing(app, interview, interviewer = {}) {
  const screening = app.screening || {};
  const interviewScore = Math.round(interview?.interviewScore || 0);
  const compositeScore = Math.round(interview?.compositeScore || interview?.finalScore || 0);
  const jdScore = Math.round(app.jdScore || screening.total_score || screening.ai_score || 0);
  const verdict = interview?.verdict || 'Review';
  const shortlistVerdict = interview?.shortlistVerdict || '';
  const strengths = screening.top_strengths || app.matchedSkills || interview?.topStrengths || [];
  const gaps = screening.key_gaps || app.missingSkills || [];
  const aiConcerns = interview?.concerns || [];
  const redFlags = screening.red_flags || [];
  const questions = buildSuggestedQuestions(app, interview);
  const feedback = interview?.aiFeedback || interview?.recommendation || '';
  const perAnswer = (interview?.perAnswerFeedback || []).slice(0, 5);
  const candidateAnswers = extractCandidateAnswers(interview);
  const jdSummary = app.jdFitSummary || screening.decision_note || screening.jd_fit_summary || '';
  const dims = dimensionRows(interview);
  const skills = app.skills || app.parsedData?.skills || screening.extracted_summary?.skills || [];
  const expYears = app.parsedData?.experience_years
    || screening.harness_profile?.total_experience_years
    || screening.extracted_summary?.experience_years;

  const html = `
    <div style="background:#fafafa;padding:16px;border-radius:8px;line-height:1.7;font-size:14px">
      <h4 style="margin:0 0 8px;color:#0D4F4F">1. Candidate snapshot</h4>
      <p>
        <strong>${esc(app.candidateName)}</strong> applied for <strong>${esc(app.jobTitle)}</strong>.
        ${expYears != null ? `Experience: ~${expYears} years.` : ''}
        ${skills.length ? `Tech stack: ${esc(skills.slice(0, 10).join(', '))}.` : ''}
        ${shortlistVerdict ? `<br><span style="color:#00B8B8">${esc(shortlistVerdict)}</span>` : ''}
      </p>

      <h4 style="margin:16px 0 8px;color:#0D4F4F">2. Resume screening highlights</h4>
      <p>Score: <strong>${jdScore}/100</strong> · Verdict: <strong>${esc(screening.verdict || app.recommendation || 'Reviewed')}</strong></p>
      ${jdSummary ? `<p style="background:#fff;padding:10px;border-radius:6px;border:1px solid #e7e5e4">${esc(jdSummary)}</p>` : ''}
      <p><strong>Strong areas</strong></p>
      ${listItems(strengths.slice(0, 5))}
      <p><strong>Areas needing verification</strong></p>
      ${listItems(gaps.slice(0, 5))}
      ${redFlags.length ? `<p><strong>Red flags</strong></p>${listItems(redFlags)}` : ''}

      <h4 style="margin:16px 0 8px;color:#0D4F4F">3. AI interview — how they passed</h4>
      <p>
        Interview: <strong>${interviewScore}/100</strong> · Verdict: <strong>${esc(verdict)}</strong><br>
        Composite: <strong>${compositeScore}/100</strong> (80% screening ${jdScore}% + 20% interview ${interviewScore}%)
      </p>
      ${dims.length ? `<p>${dims.map(([l, v]) => `${l}: ${Math.round(v)}/100`).join(' · ')}</p>` : ''}
      ${feedback ? `<p style="background:#fff;padding:10px;border-radius:6px;border:1px solid #e7e5e4"><strong>Manager summary:</strong> ${esc(feedback)}</p>` : ''}
      ${aiConcerns.length ? `<p><strong>AI concerns</strong></p>${listItems(aiConcerns)}` : ''}

      ${candidateAnswers.length ? `
        <h4 style="margin:16px 0 8px;color:#0D4F4F">Key answers from AI interview</h4>
        <ul>${candidateAnswers.map((a) => `<li style="margin-bottom:6px">${esc(String(a).slice(0, 280))}</li>`).join('')}</ul>
      ` : ''}

      <h4 style="margin:16px 0 8px;color:#0D4F4F">4. Suggested questions for ${esc(interviewer.name || 'you')}</h4>
      <ol>${questions.map((q) => `<li style="margin-bottom:8px">${esc(q)}<br><span style="color:#666;font-size:12px">What to look for: depth, specifics, trade-offs, real ownership.</span></li>`).join('')}</ol>

      ${perAnswer.length ? `
        <h4 style="margin:16px 0 8px;color:#0D4F4F">Per-answer notes from AI interview</h4>
        ${perAnswer.map((a) => `<p style="font-size:12px;margin-bottom:8px"><strong>Q:</strong> ${esc(a.question)}<br><strong>Feedback:</strong> ${esc(a.feedback)}</p>`).join('')}
      ` : ''}

      <h4 style="margin:16px 0 8px;color:#0D4F4F">5. Evaluation criteria (rate 1–5 after session)</h4>
      <ul>
        <li><strong>Technical depth</strong> — Can they explain architecture and implementation details?</li>
        <li><strong>Problem solving</strong> — Structured thinking under pressure?</li>
        <li><strong>Communication</strong> — Clear, concise explanations?</li>
        <li><strong>Culture fit</strong> — Alignment with team values and collaboration style?</li>
      </ul>
    </div>
  `;

  return { html, text: html.replace(/<[^>]+>/g, ' ') };
}

function toMlPayload(app, interview, interviewer) {
  return {
    candidate_name: app.candidateName,
    job_title: app.jobTitle,
    interviewer_name: interviewer.name || 'Interviewer',
    interviewer_role: interviewer.role || interviewer.designation || 'Panel Member',
    application: {
      jd_score: app.jdScore,
      jd_fit_summary: app.jdFitSummary,
      recommendation: app.recommendation,
      matched_skills: app.matchedSkills,
      missing_skills: app.missingSkills,
      skills: app.skills,
      parsed_data: app.parsedData,
    },
    screening: app.screening || {},
    interview: {
      interview_score: interview?.interviewScore,
      composite_score: interview?.compositeScore,
      verdict: interview?.verdict,
      shortlist_verdict: interview?.shortlistVerdict,
      recommendation: interview?.recommendation,
      ai_feedback: interview?.aiFeedback,
      concerns: interview?.concerns,
      top_strengths: interview?.topStrengths,
      technical_score: interview?.technicalScore,
      communication_score: interview?.communicationScore,
      problem_solving_score: interview?.problemSolvingScore,
      culture_fit_score: interview?.cultureFitScore,
      experience_depth_score: interview?.experienceDepthScore,
      jd_alignment_score: interview?.jdAlignmentScore,
      harness_transcript: interview?.harnessTranscript,
      answers: interview?.answers,
      qa_log: interview?.qaLog,
      per_answer_feedback: interview?.perAnswerFeedback,
    },
  };
}

async function buildInterviewerBriefing(app, interview, interviewer = {}) {
  const result = await ml.generateInterviewerBriefing(toMlPayload(app, interview, interviewer));
  if (!result?.briefing_html) {
    throw new Error('Groq interviewer briefing failed — no content returned. Check GROQ_API_KEY and ml-service.');
  }
  return { html: result.briefing_html, text: result.briefing_html, generatedBy: result.generated_by || 'groq' };
}

module.exports = {
  buildInterviewerBriefing,
  buildStaticInterviewerBriefing,
  buildSuggestedQuestions,
  extractCandidateAnswers,
};
